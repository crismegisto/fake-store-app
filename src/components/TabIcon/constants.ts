import Routes from 'constants/routes';

interface Icons {
  [name: string]: string;
}

const ICONS: Icons = {
  [Routes.HomeNavigator]: 'home',
  [Routes.Support]: 'support-agent'
};

export default ICONS;
