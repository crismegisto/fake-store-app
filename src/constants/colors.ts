export const primary = '#606c38';
export const secondary = '#283618';
export const tertiary = '#fefae0';
export const lightGray = '#C9C9C9';
