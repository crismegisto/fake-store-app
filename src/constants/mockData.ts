export const mockProducts = [
  {
    category: "men's clothing",
    id: 1,
    title: '1Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
    price: 109.95,
    rating: {
      rate: 3.9,
      count: 120
    },
    image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
    description:
      'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday'
  },
  {
    category: "men's clothing",
    id: 2,
    title: 'Mens Casual Premium Slim Fit T-Shirts',
    price: 22.3,
    rating: {
      rate: 3.9,
      count: 120
    },
    image: 'https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg',
    description:
      'Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.'
  }
];
