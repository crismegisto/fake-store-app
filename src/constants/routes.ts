enum Routes {
  HomeNavigator = 'HomeNavigator',
  Home = 'Home',
  Detail = 'Detail',
  Support = 'Support'
}

export default Routes;
