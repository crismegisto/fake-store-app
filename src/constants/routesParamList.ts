import { Product } from 'interfaces/product';

import Routes from './routes';
export type RoutesParamList = {
  [Routes.HomeNavigator]: undefined;
  [Routes.Home]: undefined;
  [Routes.Detail]: Product;
  [Routes.Support]: undefined;
};
