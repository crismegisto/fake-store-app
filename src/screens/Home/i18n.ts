import i18n from 'config/i18n';

i18n.addResources('en', 'HOME', {
  REQUEST_ERROR: 'Service out of operation'
});
