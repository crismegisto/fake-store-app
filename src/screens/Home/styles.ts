import { Platform, StyleSheet } from 'react-native';
import { tertiary } from 'src/constants/colors';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: tertiary
  },
  subContainer: {
    flex: 1,
    paddingTop: 10,
    paddingBottom: Platform.OS === 'ios' ? 60 : 90,
    paddingHorizontal: 4
  }
});

export default styles;
