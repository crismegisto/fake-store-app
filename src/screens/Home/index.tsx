import React, { useEffect, useMemo, useState } from 'react';
import { View, FlatList, ListRenderItem, SafeAreaView } from 'react-native';
import api from 'config/api';
import { ApiResponse } from 'apisauce';
import { act } from 'react-test-renderer';
import ProductCard from 'components/ProductCard';
import Header from 'components/Header';
import { Product } from 'interfaces/product';
import { ListKeyExtractor } from 'interfaces/global';

import styles from './styles';

function Home() {
  const [query, setQuery] = useState('');
  const [products, setProducts] = useState<Product[]>([]);

  useEffect(() => {
    const fetchProducts = async () => {
      const response: ApiResponse<Product[]> = await api.get('/products');
      if (response.ok) {
        act(() => {
          setProducts(response.data || []);
        });
      }
    };

    fetchProducts();
  }, []);

  const filteredProducts = useMemo<Product[]>(
    () => products.filter(({ title }) => title.toLowerCase().includes(query?.toLowerCase())),
    [products, query]
  );

  const renderItem: ListRenderItem<Product> = ({ item }) => <ProductCard {...item} />;

  const keyExtractor: ListKeyExtractor<Product> = ({ id }) => String(id);

  return (
    <SafeAreaView style={styles.container}>
      <Header query={query} onSearch={(text: string | null) => setQuery(text ?? '')} />
      <View style={styles.subContainer}>
        <FlatList
          renderItem={renderItem}
          keyExtractor={keyExtractor}
          data={query ? filteredProducts : products}
          numColumns={2}
        />
      </View>
    </SafeAreaView>
  );
}

export default Home;
