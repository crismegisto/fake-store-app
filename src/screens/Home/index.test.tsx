import React from 'react';
import { screen, render } from '@testing-library/react-native';

import { mockProducts } from 'constants/mockData';

import Home from './index';

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn()
    })
  };
});

jest.mock('@rneui/themed', () => ({
  AirbnbRating: jest.fn(),
  SearchBar: jest.fn()
}));

jest.mock('config/api', () => {
  const actualApi = jest.requireActual('config/api');
  return {
    ...actualApi,
    get: jest.fn().mockResolvedValue({ ok: true, data: mockProducts })
  };
});

beforeEach(() => {
  render(<Home />);
});

describe('Home screen Tests', () => {
  test('match snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('service success', async () => {
    expect(await screen.findAllByTestId('product-card')).toHaveLength(2);
  });
});
