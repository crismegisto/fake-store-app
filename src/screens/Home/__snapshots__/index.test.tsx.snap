// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen Tests match snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#fefae0",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "flex": 1,
        "paddingBottom": 60,
        "paddingHorizontal": 4,
        "paddingTop": 10,
      }
    }
  >
    <RCTScrollView
      data={
        [
          {
            "category": "men's clothing",
            "description": "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
            "id": 1,
            "image": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
            "price": 109.95,
            "rating": {
              "count": 120,
              "rate": 3.9,
            },
            "title": "1Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
          },
          {
            "category": "men's clothing",
            "description": "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
            "id": 2,
            "image": "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
            "price": 22.3,
            "rating": {
              "count": 120,
              "rate": 3.9,
            },
            "title": "Mens Casual Premium Slim Fit T-Shirts",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={[]}
      viewabilityConfigCallbackPairs={[]}
    >
      <View>
        <View
          onFocusCapture={[Function]}
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "white",
                  "borderRadius": 12,
                  "height": 200,
                  "margin": 6,
                  "opacity": 1,
                  "padding": 14,
                  "width": 357,
                }
              }
            >
              <Image
                resizeMode="contain"
                source={
                  {
                    "uri": "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
                  }
                }
                style={
                  {
                    "alignSelf": "center",
                    "height": 86,
                    "width": 86,
                  }
                }
              />
              <Text
                style={
                  {
                    "fontWeight": "bold",
                    "marginTop": 20,
                  }
                }
                testID="product-card"
              >
                1Fjallraven - Foldsack No. 1 Backpack, Fits 15 Lap...
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                    "position": "absolute",
                    "right": 10,
                    "top": 10,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#606c38",
                        "fontSize": 24,
                      },
                      undefined,
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  î¡½
                </Text>
              </View>
            </View>
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "backgroundColor": "white",
                  "borderRadius": 12,
                  "height": 200,
                  "margin": 6,
                  "opacity": 1,
                  "padding": 14,
                  "width": 357,
                }
              }
            >
              <Image
                resizeMode="contain"
                source={
                  {
                    "uri": "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
                  }
                }
                style={
                  {
                    "alignSelf": "center",
                    "height": 86,
                    "width": 86,
                  }
                }
              />
              <Text
                style={
                  {
                    "fontWeight": "bold",
                    "marginTop": 20,
                  }
                }
                testID="product-card"
              >
                Mens Casual Premium Slim Fit T-Shirts
              </Text>
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onClick={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "opacity": 1,
                    "position": "absolute",
                    "right": 10,
                    "top": 10,
                  }
                }
              >
                <Text
                  allowFontScaling={false}
                  selectable={false}
                  style={
                    [
                      {
                        "color": "#606c38",
                        "fontSize": 24,
                      },
                      undefined,
                      {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      {},
                    ]
                  }
                >
                  î¡½
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
